keps/sig-network/4444-service-routing-preference/README.md

```diff
@@ -0,0 +1,998 @@
+  the `internalTrafficPolicy` and `externalTrafficPolicy` fields which have a
+  more fixed behavior.
+* Use of the word "selection" for a name like `endpointSelection` were avoided
+  so as not to confuse with the actual process of selecting the complete set of
+  pods backing a service.
+
+<<[/UNRESOLVED]>>
+
+### Intersection with internal/externalTrafficPolicy
+
+The intersection of the field with `internalTrafficPolicy` and
+`externalTrafficPolicy` fields remains the same as the annotation. The following
+table borrowed from [KEP-2086: Service Internal Traffic
+Policy](https://github.com/kubernetes/enhancements/tree/master/keps/sig-network/2086-service-internal-traffic-policy)
+captures the precedence
+
+<table>
+<thead>
+  <tr>
+    <th>ExternalTrafficPolicy</th>
+    <th>InternalTrafficPolicy</th>
+    <th>routingPreference</th>
+    <th>External Result</th>
+    <th>Internal Result</th>
+  </tr>
+</thead>
+<tbody>
+  <tr>
+    <td>-</td>
+    <td>-</td>
+    <td>Auto</td>
+    <td>routingPreference=Auto</td>
+    <td>routingPreference=Auto</td>
+  </tr>
+  <tr>
+    <td>Local</td>
+    <td>-</td>
+    <td>Auto</td>
+    <td>ExternalTrafficPolicy=Local</td>
+    <td>routingPreference=Auto</td>
+  </tr>
+  <tr>
+    <td>Local</td>
+    <td>Local</td>
+    <td>Auto</td>
+    <td>ExternalTrafficPolicy=Local</td>
+    <td>InternalTrafficPolicy=Local</td>
+  </tr>
+</tbody>
+</table>
+
+### Test Plan
+
+[X] I/we understand the owners of the involved components may require updates to
+existing tests to make this code solid enough prior to committing the changes necessary
+to implement this enhancement.
+
+##### Prerequisite testing updates
+
+No updates required.
+
+##### Unit tests
+
+The main packages that will see major changes due to this enhancement are:
+- `k8s.io/kubernetes/vendor/k8s.io/endpointslice/topologycache`: `2024-01-26` -
+  `74.8`
+- `k8s.io/kubernetes/vendor/k8s.io/endpointslice`: `2024-01-26` - `80.7`
+
+There will be some code refactoring into newer packages which may prevent exact
+comparisons, but the updated packages will aim to provide equivalent coverage.
+
+The following packages will also see minor changes:
+- `k8s.io/kubernetes/pkg/controller/endpointslice`: `2024-01-26` - `61.4`
+- `k8s.io/kubernetes/pkg/proxy`: `2024-01-26` - `69.5`
+
+##### Integration tests
+
+* Verify that if both the annotation `service.kubernetes.io/topology-mode=Auto`
+  and field `routingPreference=Zone` are configured, precedence is given
+  to the annotation.
+
+##### e2e tests
+
+* Verify that EndpointSlice hints are correctly populated when
+  `routingPreference=Zone`
+* Verify through probes that when `routingPreference=Zone`, requests
+  originating from a zone which has service pods get sent to a pod in the same
+  zone. For requests originating from zones with no service pods, requests
+  should not get blackholed and should rather be forwarded to any service pod
+  from the cluster.
```

Comment:
Can we usefully end-to-end test a _preference_? It seems to me that the only thing we can actually assert is that traffic does not get dropped into a hole.